### PR TITLE
add cpucmd, which uses the cpu package

### PR DESCRIPTION
**Pull request by rminnich**
The advantage of cpucmd is that one need not install anything
on the DUT, and one need never think about copying files to the
DUT and back.

Tests do not need to be installed on the DUT; they
(and their shared libraries) are run from mountpoints back to
the host. When a test on the DUT writes files, they can be written
to mountpoints from the host. cpu manages all the mountpoints
transparently, as controlled by the CPU_NAMESPACE environment
variable. Mounts are not NFS mounts and do not require sysadmin
help or configurations in /etc. They are dynamic, i.e., can be
different for each invocation of cpu.

For a given CPU session, the mounts are private and
not visible, either on the host or via other CPU sessions.
/tmp is remounted via tmpfs, and private
They do not interfere with each other.

Finally, files on the DUT are still visible in their
usual place if they are not part of the namespace; i.e.,
files in /dev are all still visible.

Example, my host is a300, and my DUT is apu2.
rminnich@a300:~$ cpu apu2 flashrom -r xyz -p internal
...
rminnich@a300:~$ ls -l xyz
-rw-r--r-- 1 rminnich rminnich 8388608 Sep 22 12:29 xyz

The file is written to xyz on the host, not the DUT. No
scp needed to get it.

At Google, we typically have linuxboot images (i.e. in flash)
with one, and only one program, the cpu daemon. Nothing else
need be installed; we do not need working usb or even a local disk.
Just linuxboot and a cpu daemon in flash.

There's no requirement that the linuxboot+cpud be in flash; it is just
convenient. On some systems it is on a usb stuck, netbooted, or on NVME.

See github.com:linuxboot/linuxboot/pcengines/apu2 for a build example.

To build a u-root with only a cpud:
u-root -initcmd=cpud -files path/to/cpu_rsa.pub:key.pub \
	-defaultsh="" \
	github.com/u-root/cpu/cmds/cpud

In this case, cpud is pid 1.

We allow contest code to use .ssh/config to fill the port,
via the cpu GetPort function.

To run with the cpucmd plugin, you have to set CPU_NAMESPACE:
export CPU_NAMESPACE=/lib:/lib64:/usr:/bin:/etc:/home

To use binaries from a different architecture, e.g. ARM DUT
and x86 host:
export root=/path/to/arm/root
export CPU_NAMESPACE=$root/lib:$root/lib64:$root/usr:$root/bin:/etc:/home
note that the etc and home are not redirected to $root.

/home, /etc, etc. are not required. Further, if no mounts are desired,
unset CPU_NAMESPACE

In that case, however, there must be the correct binaries on the DUT.

Here is a sample .json which uses cpucmd.

{
    "JobName": "lspci test job",
    "Runs": 1,
    "RunInterval": "3s",
    "Tags": ["lscpi"],
    "TestDescriptors": [
        {
            "TargetManagerName": "TargetList",
            "TargetManagerAcquireParameters": {
                "Targets": [
                    {
                        "FQDN": "apu2",
                        "ID": "12345"
                    }
                ]
            },
            "TargetManagerReleaseParameters": {
            },
            "TestFetcherName": "literal",
            "TestFetcherFetchParameters": {
                "TestName": "Literal test",
                "Steps": [
                    {
                        "name": "cpucmd",
			"label": "CPU",
                        "parameters": {
                                "host": ["{{ .FQDN }}"],
		                "private_key_file": ["/home/rminnich/.ssh/apu2_rsa"],
				"user": ["rminnich"],
                                "executable": ["lspci"],
				"args": ["-v"],
                                "expect": ["Host Bridge"]
                        }
                    }
                ]
            }
        }
    ],
    "Reporting": {
        "RunReporters": [
            {
                "Name": "TargetSuccess",
                "Parameters": {
                    "SuccessExpression": ">80%"
                }
            }
        ]
    }
}

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

File: plugins/teststeps/cpucmd/cpucmd.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+// The cpu command uses ssh config files, e.g. ~/.ssh/config.
+// The big difference is that cpu will default to port 23, not port 22,
+// if no entries exist for a host.
+//
+// Warning: this plugin does not lock keys in memory, and does no
+// safe erase in memory to avoid forensic attacks. If you need that, please
+// submit a PR.
+//
+// Warning: commands are interpreted, so be careful with external input in the
+// test step arguments.
```
Comment (rjoleary) [inline plugins/teststeps/cpucmd/cpucmd.go:20]: The docs should go right above `package cpucmd`.

File: plugins/teststeps/cpucmd/cpucmd.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+	if false {
+		ts.User = params.GetOne("user")
+		if ts.User.IsEmpty() {
+			return errors.New("invalid or missing 'user' parameter, must be exactly one string")
+		}
+	}
+
+	ts.PrivateKeyFile = params.GetOne("private_key_file")
+	// do not fail if key file is empty, in such case it won't be used
+	ts.PrivateKeyFile = params.GetOne("private_key_file")
```
Comment (rjoleary) [inline plugins/teststeps/cpucmd/cpucmd.go:285]: I am wondering why you have this line twice.

**Review by rjoleary — APPROVED** [review]

File: plugins/teststeps/cpucmd/cpucmd.go (hunk 2 above)
Comment (rminnich) [reply plugins/teststeps/cpucmd/cpucmd.go:283]: fixed

File: plugins/teststeps/cpucmd/cpucmd.go (hunk 1 above)
Comment (rminnich) [reply plugins/teststeps/cpucmd/cpucmd.go:18]: fixed.

**Review by rjoleary — APPROVED** [review]

**Pull request merged**